diff -Naur busybox-1.31.0-orig/procps/free.c busybox-1.31.0/procps/free.c
--- busybox-1.31.0-orig/procps/free.c	2018-12-30 16:14:20.000000000 +0100
+++ busybox-1.31.0/procps/free.c	2019-08-23 13:59:43.896941300 +0200
@@ -32,9 +32,7 @@
 //usage:       "Total:       386144       257128       129016\n"
 
 #include "libbb.h"
-#ifdef __linux__
-# include <sys/sysinfo.h>
-#endif
+#include <sys/sysinfo.h>
 
 struct globals {
 	unsigned mem_unit;
diff -Naur busybox-1.31.0-orig/procps/uptime.c busybox-1.31.0/procps/uptime.c
--- busybox-1.31.0-orig/procps/uptime.c	2018-12-30 16:14:20.000000000 +0100
+++ busybox-1.31.0/procps/uptime.c	2019-08-23 14:06:03.829280300 +0200
@@ -41,9 +41,7 @@
 //usage:       "  1:55pm  up  2:30, load average: 0.09, 0.04, 0.00\n"
 
 #include "libbb.h"
-#ifdef __linux__
-# include <sys/sysinfo.h>
-#endif
+#include <sys/sysinfo.h>
 
 #ifndef FSHIFT
 # define FSHIFT 16              /* nr of bits of precision */
