diff -Naur busybox-1.31.0-orig/include/libbb.h busybox-1.31.0/include/libbb.h
--- busybox-1.31.0-orig/include/libbb.h	2019-06-10 11:09:31.000000000 +0200
+++ busybox-1.31.0/include/libbb.h	2025-05-23 17:26:38.456788400 +0200
@@ -141,6 +141,7 @@
 # include <netinet/in.h>
 #else
 # include <arpa/inet.h>
+# include <netinet/in.h>
 //This breaks on bionic:
 //# if !defined(__socklen_t_defined) && !defined(_SOCKLEN_T_DECLARED)
 ///* We #define socklen_t *after* includes, otherwise we get
diff -Naur busybox-1.31.0-orig/procps/free.c busybox-1.31.0/procps/free.c
--- busybox-1.31.0-orig/procps/free.c	2018-12-30 16:14:20.000000000 +0100
+++ busybox-1.31.0/procps/free.c	2025-05-23 17:29:13.720566800 +0200
@@ -32,9 +32,7 @@
 //usage:       "Total:       386144       257128       129016\n"
 
 #include "libbb.h"
-#ifdef __linux__
-# include <sys/sysinfo.h>
-#endif
+#include <sys/sysinfo.h>
 
 struct globals {
 	unsigned mem_unit;
diff -Naur busybox-1.31.0-orig/procps/uptime.c busybox-1.31.0/procps/uptime.c
--- busybox-1.31.0-orig/procps/uptime.c	2018-12-30 16:14:20.000000000 +0100
+++ busybox-1.31.0/procps/uptime.c	2025-05-23 17:29:31.151833100 +0200
@@ -41,9 +41,7 @@
 //usage:       "  1:55pm  up  2:30, load average: 0.09, 0.04, 0.00\n"
 
 #include "libbb.h"
-#ifdef __linux__
-# include <sys/sysinfo.h>
-#endif
+#include <sys/sysinfo.h>
 
 #ifndef FSHIFT
 # define FSHIFT 16              /* nr of bits of precision */
