diff -Naur libdrm-2.4.100-orig/include/drm/drm.h libdrm-2.4.100/include/drm/drm.h
--- libdrm-2.4.100-orig/include/drm/drm.h	2019-10-16 23:36:48.000000000 +0200
+++ libdrm-2.4.100/include/drm/drm.h	2020-10-16 19:27:00.829357200 +0200
@@ -36,7 +36,7 @@
 #ifndef _DRM_H_
 #define _DRM_H_
 
-#if   defined(__linux__)
+#if   defined(__linux__) || defined(__KOS__)
 
 #include <linux/types.h>
 #include <asm/ioctl.h>
@@ -806,7 +806,9 @@
 }
 #endif
 
+#ifndef _DRM_MODE_H
 #include "drm_mode.h"
+#endif /* !_DRM_MODE_H */
 
 #if defined(__cplusplus)
 extern "C" {
diff -Naur libdrm-2.4.100-orig/include/drm/drm_mode.h libdrm-2.4.100/include/drm/drm_mode.h
--- libdrm-2.4.100-orig/include/drm/drm_mode.h	2019-10-16 23:36:48.000000000 +0200
+++ libdrm-2.4.100/include/drm/drm_mode.h	2020-10-16 19:27:09.783495400 +0200
@@ -27,7 +27,9 @@
 #ifndef _DRM_MODE_H
 #define _DRM_MODE_H
 
+#ifndef _DRM_H_
 #include "drm.h"
+#endif /* !_DRM_H_ */
 
 #if defined(__cplusplus)
 extern "C" {
diff -Naur libdrm-2.4.100-orig/include/drm/nouveau_drm.h libdrm-2.4.100/include/drm/nouveau_drm.h
--- libdrm-2.4.100-orig/include/drm/nouveau_drm.h	2019-10-16 23:36:48.000000000 +0200
+++ libdrm-2.4.100/include/drm/nouveau_drm.h	2020-10-16 19:27:26.079300100 +0200
@@ -58,7 +58,11 @@
 struct drm_nouveau_grobj_alloc {
 	int      channel;
 	uint32_t handle;
+#ifdef __cplusplus
+	int      class_;
+#else /* __cplusplus */
 	int      class;
+#endif /* !__cplusplus */
 };
 
 struct drm_nouveau_notifierobj_alloc {
diff -Naur libdrm-2.4.100-orig/include/drm/via_drm.h libdrm-2.4.100/include/drm/via_drm.h
--- libdrm-2.4.100-orig/include/drm/via_drm.h	2019-10-16 23:36:48.000000000 +0200
+++ libdrm-2.4.100/include/drm/via_drm.h	2020-10-16 19:27:38.626868400 +0200
@@ -37,7 +37,14 @@
 #ifndef _VIA_DEFINES_
 #define _VIA_DEFINES_
 
+#ifndef __has_include
+#define __NO_has_include 1
+#define __has_include(x) 0
+#endif /* !__has_include */
+
+#if __has_include("via_drmclient.h")
 #include "via_drmclient.h"
+#endif /* __has_include("via_drmclient.h") */
 
 #define VIA_NR_SAREA_CLIPRECTS		8
 #define VIA_NR_XVMC_PORTS               10
diff -Naur libdrm-2.4.100-orig/xf86drm.h libdrm-2.4.100/xf86drm.h
--- libdrm-2.4.100-orig/xf86drm.h	2019-10-16 23:36:48.000000000 +0200
+++ libdrm-2.4.100/xf86drm.h	2020-10-16 19:32:02.939744900 +0200
@@ -47,7 +47,7 @@
 #define DRM_MAX_MINOR   16
 #endif
 
-#if defined(__linux__)
+#if defined(__linux__) || defined(__KOS__)
 
 #define DRM_IOCTL_NR(n)		_IOC_NR(n)
 #define DRM_IOC_VOID		_IOC_NONE
