diff -Naur xproto-7.0.23-orig/Xarch.h xproto-7.0.23/Xarch.h
--- xproto-7.0.23-orig/Xarch.h	2012-03-16 07:11:44.000000000 +0100
+++ xproto-7.0.23/Xarch.h	2020-08-13 15:48:43.177429200 +0200
@@ -47,7 +47,7 @@
 #    include <sys/types.h>
 #   endif
 #   include <machine/endian.h>
-#  elif defined(linux)
+#  elif defined(linux) || defined(__KOS__)
 #   if defined __STRICT_ANSI__
 #    undef __STRICT_ANSI__
 #    include <endian.h>
diff -Naur xproto-7.0.23-orig/Xthreads.h xproto-7.0.23/Xthreads.h
--- xproto-7.0.23-orig/Xthreads.h	2012-03-16 07:11:44.000000000 +0100
+++ xproto-7.0.23/Xthreads.h	2020-08-11 09:53:07.369682500 +0200
@@ -259,11 +259,11 @@
 #       define xthread_clear_id(id) id = _X_no_thread_id
 #       define xthread_equal(id1,id2) pthread_equal(id1, id2)
 #      endif /* _DECTHREADS_ */
-#      if defined(__linux__)
+#      if defined(__linux__) || defined(__KOS__)
 #       define xthread_have_id(id) !pthread_equal(id, 0)
 #       define xthread_clear_id(id) id = 0
 #       define xthread_equal(id1,id2) pthread_equal(id1, id2)
-#      endif /* linux */
+#      endif /* __linux__ || __KOS__ */
 #      if defined(_CMA_VENDOR_) && defined(_CMA__IBM) && (_CMA_VENDOR_ == _CMA__IBM)
 #       ifdef DEBUG		/* too much of a hack to enable normally */
 /* see also cma__obj_set_name() */
