diff -Naur xtrans-1.4.0-orig/Xtranssock.c xtrans-1.4.0/Xtranssock.c
--- xtrans-1.4.0-orig/Xtranssock.c	2020-10-16 13:55:33.010759500 +0200
+++ xtrans-1.4.0/Xtranssock.c	2020-10-16 13:55:40.094890500 +0200
@@ -396,7 +396,11 @@
 
     if ((ciptr->fd = socket(Sockettrans2devtab[i].family, type,
 	Sockettrans2devtab[i].protocol)) < 0
-#ifndef WIN32
+#if !defined(WIN32) && !defined(__KOS__)
+       /* On KOS, sysconf(_SC_OPEN_MAX) == -1  (which posix defines as 'unlimited')
+        * So really, this code right here used to check for `fd < 0 || fd >= -1',
+        * which would obviously always result in it thinking that the socket()
+        * call failed... UGH!!! */
 #if (defined(X11_t) && !defined(USE_POLL)) || defined(FS_t) || defined(FONT_t)
        || ciptr->fd >= sysconf(_SC_OPEN_MAX)
 #endif
